zkevm-circuits/src/evm_circuit/execution/exp.rs

```diff
@@ -0,0 +1,301 @@
+                );
+                // lookup for last step, i.e. (is_first == 0, is_last, base, 2, base^2)
+                cb.exp_table_lookup(
+                    identifier,
+                    single_step.expr(),
+                    1.expr(),
+                    base_limbs,
+                    [2.expr(), 0.expr()], // exponent == 2
+                    2.expr(),             // lsb_exponent == 2
+                    [base_sq_lo.expr(), base_sq_hi.expr()],
+                );
+            },
+        );
+
+        // In order to calculate the dynamic gas cost of the exponentiation operation,
+        // we need the byte-size of the exponent, i.e. the minimum number of
+        // bytes that can represent the exponent value.
+        let exponent_byte_size = ByteSizeGadget::construct(cb, &exponent_rlc);
+
+        // Finally we build an expression for the dynamic gas cost.
+        let dynamic_gas_cost = 50.expr() * exponent_byte_size.byte_size();
+        let step_state_transition = StepStateTransition {
+            rw_counter: Transition::Delta(3.expr()),
+            program_counter: Transition::Delta(1.expr()), // 2 stack pops, 1 stack push
+            stack_pointer: Transition::Delta(1.expr()),
+            gas_left: Transition::Delta(
+                -OpcodeId::EXP.constant_gas_cost().expr() - dynamic_gas_cost,
+            ),
+            ..Default::default()
+        };
+        let same_context = SameContextGadget::construct(cb, opcode, step_state_transition);
+
+        Self {
+            same_context,
+            base: base_rlc,
+            base_sq,
+            zero_rlc,
+            base_sq_mul_gadget,
+            exponent: exponent_rlc,
+            exponentiation: exponentiation_rlc,
+            exponent_is_zero,
+            exponent_is_one,
+            single_step,
+            exponent_byte_size,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _tx: &Transaction,
+        _call: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        let [base, exponent, exponentiation] =
+            [step.rw_indices[0], step.rw_indices[1], step.rw_indices[2]]
+                .map(|idx| block.rws[idx].stack_value());
+
+        self.base.assign(region, offset, Some(base.to_le_bytes()))?;
+        self.exponent
+            .assign(region, offset, Some(exponent.to_le_bytes()))?;
+        self.exponentiation
+            .assign(region, offset, Some(exponentiation.to_le_bytes()))?;
+
+        let exponent_scalar = exponent
+            .to_scalar()
+            .expect("exponent should fit into scalar");
+        self.exponent_is_zero
+            .assign(region, offset, exponent_scalar)?;
+        self.exponent_is_one
+            .assign(region, offset, exponent_scalar, F::one())?;
```

Comment:
these witness assignments will fail in the case `exponent_scalar = 2^129` (emit verify failure error).